### ISSUE TITLE
[API Proposal]: Add JsonElement.ValueEquals(JsonElement)

### ISSUE THREAD

**Issue by bill-poole** [dotnet/runtime#99994]
### Background and motivation

The `System.Text.Json.JsonElement` struct currently provides `ValueEquals` overloads for comparing the value of a `JsonElement` with a string (`string`, `ReadOnlySpan<char>` or `ReadOnlySpan<byte>`) representation of another `JsonElement` instance. However, if we want to compare two `JsonElement` instances, then this requires the following, which requires heap allocations and UTF-8 decoding.

```csharp
jsonElement1.ValueEquals(jsonElement2.GetString());
```

It would be presumably much more efficient to be able to do:

```csharp
jsonElement1.ValueEquals(jsonElement2);
```

### API Proposal

```csharp
namespace System.Text.Json;

public readonly struct JsonElement
{
    public bool ValueEquals(JsonElement other);
}
```

### API Usage

```csharp
// Compare jsonElement1 with jsonElement2
Console.WriteLine($"Values are equal: {jsonElement1.ValueEquals(jsonElement2)}");
```

**Comment by gregsdennis** [comment]
I have this as an extension in _Json.More.Net_.

```c#
var isEqual = elt1.IsEquivalentTo(elt2);
```

The code is [here](https://github.com/gregsdennis/json-everything/blob/master/Json.More%2FJsonElementExtensions.cs) if you want to inspect.

**Comment by bill-poole** [comment]
Thanks @gregsdennis! 

Your implementation still allocates though if the two `JsonElement` instances are `JsonValueKind.String`, `JsonValueKind.Array` or `JsonValueKind.Object`. I'm hoping a `JsonElement.ValueEquals(JsonElement)` overload would be allocation-free.

Also, comparing two `JsonElement` instances of type `JsonValueKind.Number` by parsing them both into `Decimal` instances is lossy if the two numbers have more decimal places than supported by the `System.Decimal` type.

**Comment by gregsdennis** [comment]
> Also, comparing two `JsonElement` instances of type `JsonValueKind.Number` by parsing them both into `Decimal` instances is lossy if the two numbers have more decimal places than supported by the `System.Decimal` type.

This is true regardless of what .Net numeric type you use.  It's an inherent consequence of a limited practical numeric space vs JSON's unlimited theoretical numeric space.

I chose `decimal` because I think my clients would prefer the extra precision over the extra range of `double`.

**Comment by bill-poole** [comment]
>This is true regardless of what .Net numeric type you use.

Yes, but I would imagine a `JsonElement.ValueEquals(JsonElement)` overload would compare the numbers as strings, rather than converting them into a binary floating point type with limited precision/range.

**Comment by gregsdennis** [comment]
Yeah, that doesn't work, either: #97490.

**Comment by bill-poole** [comment]
But isn't that the current behavior of `JsonElement.ValueEquals(String)`, `JsonElement.ValueEquals(ReadOnlySpan<char>)` and `JsonElement.ValueEquals(ReadOnlySpan<byte>)`? I looked at their implementations and they don't deserialize then compare, they just compare the strings.

I would therefore propose that `JsonElement.ValueEquals(JsonElement)` would behave the exact same way. i.e., `jsonElement1.ValueEquals(jsonElement2)` would do exactly the same thing as `jsonElement1.ValueEquals(jsonElement2.GetString())`, but without the unnecessary allocation/decode.

**Comment by gregsdennis** [comment]
@eiriktsarpalis I think the ability to do this without allocation on `JsonElement` would be valuable.  I don't think this is the same as that issue you linked.

**Comment by bill-poole** [comment]
This now seems to be possible using either `JsonElement.DeepEquals(JsonElement)` or `JsonMarshal.GetRawUtf8Value(JsonElement)` added in .NET 9, depending on whether we want to normalize number formats, spacing, etc.

```csharp
public static class JsonElementExtensions
{
    public static bool ValueEquals(this JsonElement element, JsonElement other, bool normalize)
        => normalize ? JsonElement.DeepEquals(element, other)
            : JsonMarshal.GetRawUtf8Value(element).SequenceEqual(JsonMarshal.GetRawUtf8Value(other));
}
```

**Comment by eiriktsarpalis** [comment]
@bill-poole why would you not want to normalize though? Any kind of whitespace would be sufficient to trigger false negatives.

**Comment by bill-poole** [comment]
If you’re comparing elements of documents that are guaranteed to have been created internally by an application such that the formatting is guaranteed to be already normalized, then just comparing the UTF-8 is sufficient and more efficient.
